src/lib.rs

```diff
@@ -134,13 +137,129 @@
 //! }
 //! ```
 
+#![no_std]
+
 extern crate alloc;
 
 use alloc::boxed::Box;
+use alloc::rc::{Rc, Weak as RcWeak};
+use alloc::sync::{Arc, Weak as ArcWeak};
+use alloc::vec::Vec;
 use core::fmt;
 use core::mem::{self, MaybeUninit};
 use core::ops::{Deref, DerefMut};
-use core::ptr;
+use core::ptr::{self, NonNull};
+
+/// Marker trait to indicate which types are safe to stow.
+///
+/// Currently, it is undefined behavior to read/write uninitialized memory
+/// in a primitive type, such as a pointer or usize, even if that memory is
+/// never "used". This means that types that have uninitialized bytes in their
+/// representation (such as structs with padding bytes) cannot be packed by
+/// stowaway, even if they would otherwise fit. It's not currently possible
+/// to detect if a struct might contain uninitialized bytes, so any types that
+/// want to be stowed must implement this trait.
+///
+/// This trait is implemented for most primitive and standard library types
+/// you might want to stow:
+/// - All integer & float types.
+/// - `bool` and `char`.
+/// - All primitive pointer types.
+/// - All arrays of `Stowable` (up to length 32, until const generics are stable)
+/// - All structured pointer types (`Rc`, `Arc`, etc)
+///
+/// # Safety
+///
+/// Correct usage of this trait is critical to stowaway's safety guarantees.
+/// This struct can only be implemented on types that can never have
+/// uninitialized bytes *anywhere* in their representation (including padding
+/// bytes). Making this determination isn't trivial, though typically it's
+/// safe if `sizeof(T) == sum(sizeof(fields in T))`, and all of the fields in
+/// `T` are themselves `Stowable`.
+///
+/// Structs which are `repr(transparent)` are `Stowable` if their inner field
+/// is `Stowable`.
+pub unsafe trait Stowable: Sized {}
```

Comment:
I don't think we need to enforce `Sized` here, it prevents other people from using `Stowable` as a super-trait if they want dynamic dispatch, and `Stowaway` already requires `T: Sized` by default.